Models/MCTS_torch.py

```diff
@@ -0,0 +1,415 @@
+            for ix, iy, idx in zip(hidden_state_index_x_lst, hidden_state_index_y_lst, range(num_states)):
+                tensors_states[idx] = hidden_state_pool[ix][iy]
+
+            # Inside the search tree we use the dynamics function to obtain the next
+            # hidden state given an action and the previous hidden state.
+
+            last_action = np.asarray(last_action)
+
+            # 0.026 to 0.064 seconds
+            # self.ckpt_time = time.time_ns()
+            network_output = self.network.recurrent_inference(tensors_states, last_action)
+            # print("recurrent inference took: {}".format(time.time_ns() - self.ckpt_time))
+
+            value_prefix_pool = np.array(network_output["value_logits"]).reshape(-1).tolist()
+            value_pool = np.array(network_output["value"]).reshape(-1).tolist()
+
+            # 0.002 seconds
+            policy_logits, _, mappings, _ = self.sample(network_output["policy_logits"].numpy(),
+                                                        self.default_string_mapping, self.default_byte_mapping,
+                                                        config.NUM_SAMPLES)
+            # These assignments take 0.0001 > time
+            # add nodes to the pool after each search
+            hidden_states_nodes = network_output["hidden_state"]
+            hidden_state_pool.append(hidden_states_nodes)
+
+            reset_idx = (np.array(search_lens) % horizons == 0)
+            is_reset_lst = reset_idx.astype(np.int32).tolist()
+            # tree node.isreset = is_reset_list[node]
+            hidden_state_index_x += 1
+
+            # 0.001 seconds
+            # backpropagation along the search path to update the attributes
+            tree.batch_back_propagate(hidden_state_index_x, discount, value_prefix_pool, value_pool, policy_logits,
+                                      min_max_stats_lst, results, is_reset_lst, mappings)
+
+    @staticmethod
+    def select_action(visit_counts, temperature=1, deterministic=True):
+        """select action from the root visit counts.
+        Parameters
+        ----------
+        visit_counts: list
+            visit counts for each child
+        temperature: float
+            the temperature for the distribution
+        deterministic: bool
+            True -> select the argmax
+            False -> sample from the distribution
+        """
+        action_probs = [visit_count_i ** (1 / temperature) for visit_count_i in visit_counts]
+        total_count = sum(action_probs)
+        action_probs = [x / total_count for x in action_probs]
+        if deterministic:
+            action_pos = np.argmax([v for v in visit_counts])
+        else:
+            action_pos = np.random.choice(len(visit_counts), p=action_probs)
+
+        count_entropy = entropy(action_probs, base=2)
+        return action_pos, count_entropy
+
+    @staticmethod
+    def encode_action_to_str(policy_logits, mask):
+        actions = []
```

Comment:
This function changed on main a bit, when the merge happens, you will simply need to copy.